File: Day_1/code/04_Implemented_trie_2.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.wordCount = 0
        self.prefixCount = 0

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
            node.prefixCount += 1
        node.wordCount += 1

    def countWordsEqualTo(self, word: str) -> int:
        node = self.root
        for ch in word:
            if ch not in node.children:
                return 0
            node = node.children[ch]
        return node.wordCount

    def countWordsStartingWith(self, prefix: str) -> int:
        node = self.root
        for ch in prefix:
            if ch not in node.children:
                return 0
            node = node.children[ch]
        return node.prefixCount

    def erase(self, word: str) -> None:
        node = self.root
        stack = []
        for ch in word:
            stack.append((node, ch))
            node = node.children[ch]
            node.prefixCount -= 1
        node.wordCount -= 1


trie = Trie()
trie.insert("apple")
trie.insert("apple")
print(trie.countWordsEqualTo("apple"))     # 2
print(trie.countWordsStartingWith("app"))  # 2
trie.erase("apple")
print(trie.countWordsEqualTo("apple"))     # 1
print(trie.countWordsStartingWith("app"))  # 1

File: Day_1/code/04_Implemented_trie_2.java
import java.util.HashMap;

class TrieNode {
    HashMap<Character, TrieNode> children;
    int wordCount;
    int prefixCount;

    TrieNode() {
        children = new HashMap<>();
        wordCount = 0;
        prefixCount = 0;
    }
}

public class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode node = root;
        for (char ch : word.toCharArray()) {
            node.children.putIfAbsent(ch, new TrieNode());
            node = node.children.get(ch);
            node.prefixCount++;
        }
        node.wordCount++;
    }

    public int countWordsEqualTo(String word) {
        TrieNode node = root;
        for (char ch : word.toCharArray()) {
            if (!node.children.containsKey(ch)) return 0;
            node = node.children.get(ch);
        }
        return node.wordCount;
    }

    public int countWordsStartingWith(String prefix) {
        TrieNode node = root;
        for (char ch : prefix.toCharArray()) {
            if (!node.children.containsKey(ch)) return 0;
            node = node.children.get(ch);
        }
        return node.prefixCount;
    }

    public void erase(String word) {
        TrieNode node = root;
        for (char ch : word.toCharArray()) {
            TrieNode nextNode = node.children.get(ch);
            nextNode.prefixCount--;
            node = nextNode;
        }
        node.wordCount--;
    }
}


// Java
Trie trie = new Trie();
trie.insert("apple");
trie.insert("apple");
System.out.println(trie.countWordsEqualTo("apple")); // 2
trie.erase("apple");
System.out.println(trie.countWordsEqualTo("apple")); // 1
System.out.println(trie.countWordsStartingWith("app")); // 1
